File: validate.test.ts
import { describe, it, expect } from "bun:test";
import {
  validate,
  DiffResult,
  VerifierSpec,
  ValidationResult,
} from "./validate";

function runValidate(verifier: VerifierSpec, diffs: DiffResult[]): string[] {
  const results: string[] = [];
  // Patch validate to push to results instead of console.log
  const origLog = console.log;
  console.log = (msg: string) => results.push(msg);
  try {
    validate(verifier, diffs);
  } finally {
    console.log = origLog;
  }
  return results;
}

describe("validate", () => {
  it("matches exact insert", () => {
    const verifier: VerifierSpec = {
      type: "state_mutation_match",
      mutations: [
        {
          action: "INSERT",
          tablename: "foo",
          values: { id: 1, name: "bar" },
        },
      ],
    };
    const diffs: DiffResult[] = [
      { table: "foo", method: "insert", record: { id: 1, name: "bar" } },
    ];
    const results = validate(verifier, diffs);
    expect(results[0].success).toBe(true);
    expect(results[0].type).toBe("INSERT");
    expect(results[0].expected).toEqual(verifier.mutations[0]);
    expect(results[0].actual).toEqual(diffs[0]);
  });

  it("fails on missing mutation", () => {
    const verifier: VerifierSpec = {
      type: "state_mutation_match",
      mutations: [
        {
          action: "INSERT",
          tablename: "foo",
          values: { id: 1, name: "bar" },
        },
      ],
    };
    const diffs: DiffResult[] = [];
    const results = validate(verifier, diffs);
    expect(results[0].success).toBe(false);
    expect(results[0].actual).toBe(null);
    expect(results[0].expected).toEqual(verifier.mutations[0]);
    expect(results[0].type).toBe("INSERT");
  });

  it("matches regex", () => {
    const verifier: VerifierSpec = {
      type: "state_mutation_match",
      mutations: [
        {
          action: "INSERT",
          tablename: "foo",
          values: { name: { type: "regex", regex: ".*bar.*" } },
        },
      ],
    };
    const diffs: DiffResult[] = [
      { table: "foo", method: "insert", record: { name: "hello bar world" } },
    ];
    const results = validate(verifier, diffs);
    expect(results[0].success).toBe(true);
    expect(results[0].actual).toEqual(diffs[0]);
    expect(results[0].expected).toEqual(verifier.mutations[0]);
    expect(results[0].type).toBe("INSERT");
  });

  it("matches mutation_variable", () => {
    const verifier: VerifierSpec = {
      type: "state_mutation_match",
      mutations: [
        {
          action: "INSERT",
          tablename: "foo",
          values: {
            id: { type: "mutation_variable", name: "id" },
            name: "bar",
          },
        },
      ],
    };
    const diffs: DiffResult[] = [
      { table: "foo", method: "insert", record: { id: 123, name: "bar" } },
    ];
    const results = validate(verifier, diffs);
    expect(results[0].success).toBe(true);
    expect(results[0].actual).toEqual(diffs[0]);
    expect(results[0].expected).toEqual(verifier.mutations[0]);
    expect(results[0].type).toBe("INSERT");
  });

  it("matches semantic_match_variable (placeholder)", () => {
    const verifier: VerifierSpec = {
      type: "state_mutation_match",
      mutations: [
        {
          action: "INSERT",
          tablename: "foo",
          values: {
            name: {
              type: "semantic_match_variable",
              description: "should be bar",
            },
          },
        },
      ],
    };
    const diffs: DiffResult[] = [
      { table: "foo", method: "insert", record: { name: "bar" } },
    ];
    const results = validate(verifier, diffs);
    expect(results[0].success).toBe(true);
    expect(results[0].actual).toEqual(diffs[0]);
    expect(results[0].expected).toEqual(verifier.mutations[0]);
    expect(results[0].type).toBe("INSERT");
  });

  it("matches update with where", () => {
    const verifier: VerifierSpec = {
      type: "state_mutation_match",
      mutations: [
        {
          action: "UPDATE",
          tablename: "foo",
          where: { id: 1 },
          values: { name: "baz" },
        },
      ],
    };
    const diffs: DiffResult[] = [
      {
        table: "foo",
        method: "update",
        where: { id: 1 },
        record: { name: "baz" },
      },
    ];
    const results = validate(verifier, diffs);
    expect(results[0].success).toBe(true);
    expect(results[0].actual).toEqual(diffs[0]);
    expect(results[0].expected).toEqual(verifier.mutations[0]);
    expect(results[0].type).toBe("UPDATE");
  });
});

File: validate_test.py
import pytest
import sys
import io
from unittest.mock import patch
from typing import List, Dict, Any

from validate import (
    validate,
    DiffResult,
    VerifierSpec,
    ValidationResult,
    StateMutationMatchVerifier,
    MutationVerifier,
    DiffInsert,
    DiffUpdate,
    DiffDelete
)


def run_validate(verifier: VerifierSpec, diffs: List[DiffResult]) -> List[str]:
    """Run validate and capture output instead of printing"""
    results = []
    
    # Capture stdout
    captured_output = io.StringIO()
    with patch('sys.stdout', captured_output):
        # We need to modify the validate function to return results instead of printing
        # For now, we'll just call the validate function and return the results
        validation_results = validate(verifier, diffs)
        
        # Convert results to string format similar to the original
        for r in validation_results:
            if r.success:
                results.append(f"✔ Matched mutation: {r.type} on {r.expected.tablename}")
            else:
                results.append(f"✘ No match for mutation: {r.type} on {r.expected.tablename}")
    
    return results


class TestValidate:
    def test_matches_exact_insert(self):
        verifier = StateMutationMatchVerifier(
            type="state_mutation_match",
            mutations=[
                MutationVerifier(
                    action="INSERT",
                    tablename="foo",
                    values={"id": 1, "name": "bar"}
                )
            ]
        )
        
        diffs = [
            DiffInsert(table="foo", method="insert", record={"id": 1, "name": "bar"})
        ]
        
        results = validate(verifier, diffs)
        
        assert len(results) == 1
        assert results[0].success is True
        assert results[0].type == "INSERT"
        assert results[0].expected == verifier.mutations[0]
        assert results[0].actual == diffs[0]

    def test_fails_on_missing_mutation(self):
        verifier = StateMutationMatchVerifier(
            type="state_mutation_match",
            mutations=[
                MutationVerifier(
                    action="INSERT",
                    tablename="foo",
                    values={"id": 1, "name": "bar"}
                )
            ]
        )
        
        diffs = []
        
        results = validate(verifier, diffs)
        
        assert len(results) == 1
        assert results[0].success is False
        assert results[0].actual is None
        assert results[0].expected == verifier.mutations[0]
        assert results[0].type == "INSERT"

    def test_matches_regex(self):
        verifier = StateMutationMatchVerifier(
            type="state_mutation_match",
            mutations=[
                MutationVerifier(
                    action="INSERT",
                    tablename="foo",
                    values={"name": {"type": "regex", "regex": ".*bar.*"}}
                )
            ]
        )
        
        diffs = [
            DiffInsert(table="foo", method="insert", record={"name": "hello bar world"})
        ]
        
        results = validate(verifier, diffs)
        
        assert len(results) == 1
        assert results[0].success is True
        assert results[0].actual == diffs[0]
        assert results[0].expected == verifier.mutations[0]
        assert results[0].type == "INSERT"

    def test_matches_mutation_variable(self):
        verifier = StateMutationMatchVerifier(
            type="state_mutation_match",
            mutations=[
                MutationVerifier(
                    action="INSERT",
                    tablename="foo",
                    values={
                        "id": {"type": "mutation_variable", "name": "id"},
                        "name": "bar"
                    }
                )
            ]
        )
        
        diffs = [
            DiffInsert(table="foo", method="insert", record={"id": 123, "name": "bar"})
        ]
        
        results = validate(verifier, diffs)
        
        assert len(results) == 1
        assert results[0].success is True
        assert results[0].actual == diffs[0]
        assert results[0].expected == verifier.mutations[0]
        assert results[0].type == "INSERT"

    def test_matches_semantic_match_variable_placeholder(self):
        verifier = StateMutationMatchVerifier(
            type="state_mutation_match",
            mutations=[
                MutationVerifier(
                    action="INSERT",
                    tablename="foo",
                    values={
                        "name": {
                            "type": "semantic_match_variable",
                            "description": "should be bar"
                        }
                    }
                )
            ]
        )
        
        diffs = [
            DiffInsert(table="foo", method="insert", record={"name": "bar"})
        ]
        
        results = validate(verifier, diffs)
        
        assert len(results) == 1
        assert results[0].success is True
        assert results[0].actual == diffs[0]
        assert results[0].expected == verifier.mutations[0]
        assert results[0].type == "INSERT"

    def test_matches_update_with_where(self):
        verifier = StateMutationMatchVerifier(
            type="state_mutation_match",
            mutations=[
                MutationVerifier(
                    action="UPDATE",
                    tablename="foo",
                    where={"id": 1},
                    values={"name": "baz"}
                )
            ]
        )
        
        diffs = [
            DiffUpdate(
                table="foo",
                method="update",
                where={"id": 1},
                record={"name": "baz"}
            )
        ]
        
        results = validate(verifier, diffs)
        
        assert len(results) == 1
        assert results[0].success is True
        assert results[0].actual == diffs[0]
        assert results[0].expected == verifier.mutations[0]
        assert results[0].type == "UPDATE"


if __name__ == "__main__":
    pytest.main([__file__]) 